lib/travis/build/bash/travis_terminate.bash

```diff
@@ -7,6 +7,17 @@ travis_terminate() {
   "_travis_terminate_${TRAVIS_OS_NAME}" "${@}"
 }
 
+exit() {
+  local code="${1:-$?}"
+  # Should work as normal `exit` in subshells of the main shell.
+  # In unrelated shells, this code won't be imported.
+  if [ $$ -ne $BASHPID ]; then
+    _travis_terminate_exit "$code"
+  else
+    travis_terminate "$code"
+  fi
+}
```

Comment:
Would it perhaps be possible (and simpler?) to use `trap travis_handle_exit EXIT` at the start of the main shell script, with `travis_handle_exit` a suitable script function that e.g. just invokes `travis_terminate` or so?

Reply:
It looks so!

Tests show that

* The `EXIT` trap is not called upon exiting a subshell, saving us from having to handle that case.
* It's inherited into subshells in `bash` and not inherited in `zsh` (OSX), but since it's not called upon exiting subshells, that doesn't make a difference.